### ISSUE TITLE
Pyrsistent

### ISSUE THREAD

**Issue by gilch** [hylang/hy#1220]
One of the nice things about Clojure is the persistent immutable data structures. These have been implemented in Python via the [Pyrsistent](https://github.com/tobgu/pyrsistent) library (and perhaps others).

It might be nice if we could use some of these with a short reader macro. It seems like a common desire for those coming from Clojure. Rather than letting everyone implement their own macro versions, this might be worth putting in contrib.

A more radical step would be to make Hy's default data structures use the Pyrsistent versions, and make the Python native structures available via the reader macros. 

I'm less comfortable with that idea. It would create a dependency and might be less compatible with Python libraries or potentially future Python syntax. Though since Python is duck typed, probably not _that_ much less compatible. Also, Pyrsistent is MIT licensed, so it wouldn't be that much of a problem as a dependency.

**Comment by kirbyfan64** [comment]
+1 to the idea of using reader macros, -1 to changing Hy's defaults. Part of the awesomeness of Hy comes from the fact that it's ultimately still Python, and changing that would partly lose that benefit, along with slowing things down quite a bit.

**Comment by Kodiologist** [comment]
Yeah, adding some syntactic sugar for these things to hy.contrib seems fair enough, but making `[1 2 3]` create something other than a plain `list` by default is asking for trouble.

**Comment by gilch** [comment]
Something as simple as
```Hy
(import pyrsistent)

(defreader p [form]
  `(pyrsistent.freeze ~form) )
```
would get us pretty far. That's enough for `PVector` `PMap` and `PSet`. The `freeze` function is recursive, so it also works on nested structures,
```Hy
=> #p[#{1 2} {1 "a"  2 "b"}]
pyrsistent.freeze([{1, 2}, {1: 'a', 2: 'b', }])
pvector([pset([1, 2]), pmap({1: 'a', 2: 'b'})])
```
but there are limitations:
```Hy
=> #p #{[1 2] {1 "a"  2 "b"}}
pyrsistent.freeze({[1, 2], {1: 'a', 2: 'b', }})
Traceback (most recent call last):
  File "<input>", line 1, in <module>
TypeError: unhashable type: 'dict'
=> #p #{#p[1 2] #p{1 "a"  2 "b"}}
pyrsistent.freeze({pyrsistent.freeze([1, 2]), pyrsistent.freeze({1: 'a', 2: 'b', })})
pset([pmap({1: 'a', 2: 'b'}), pvector([1, 2])])
```
This may or may not be what we want.

There's a lot more to Pyrsistent than `freeze` though.

**Comment by Kodiologist** [comment]
Although, `PVector` and `PSet` seem equivalent to Python's built-in tuples and frozensets, respectively.

**Comment by gilch** [comment]
The interface seems basically the same, but I think PVector/PSet share data between revisions in a [trie](https://en.wikipedia.org/wiki/Trie) like Clojure does. I'm pretty sure Python's tuple and frozenset implementations don't do that.

**Comment by gilch** [comment]
Perhaps we should put this in extra with an optional dependency for Pyrsistent. This would keep Hy's core install smaller.
